### ISSUE TITLE
Add text scrolling and font changing

### ISSUE THREAD

**Issue by pascaltuf** [JerwuQu/ggoled#4]
Hi, thanks again for what you created, thanks to you I created some scripts that allow me to view the time and music playing on Spotify on the GameDac.
I created [my own repository](https://github.com/pascaltuf/ggoled-script) with these scripts if you want to take a look to test them I would be happy.
On the script I made sure that if the text is too long it divides it into parts, so as to bypass the text scrolling (in case it is difficult for you to add it)
I also tried to add the hardware code of the keyboard so as to add more supported devices to this program also changing the resolution but I did not succeed, do you have any idea why? (I'm talking about [Apex Pro](https://steelseries.com/gaming-keyboards/apex-pro))
As for the new features I saw the TODO you added and I would like you to add the ability to change the text font and the text scrolling.

**Comment by pascaltuf** [comment]
I forgot a problem that I don't know if you are aware of regarding the display, the display can support a maximum of 18 characters but for some reason from the second line onwards it is moved by one character by removing the eighteenth character as you see in the photo and therefore it is not aligned, do you know why?
I tested it by running this command:
ggoled text "ABCDEFGHIJKLMNOPQRS `n ABCDEFGHIJKLMNOPQRS `n ABCDEFGHIJKLMNOPQRS `n ABCDEFGHIJKLMNOPQRS"

![1](https://github.com/user-attachments/assets/6a7c0678-9ca6-4cf0-9e59-a42f9ba8919c)


Also if you run that command again but with a different text the text is written from above the other text, the only way is to do a "clear" and then it works as expected.

![2](https://github.com/user-attachments/assets/914426cd-206e-45a4-b0a4-7c69181cf7f1)

Instead if I add the commands to center the text (ggoled text "TEST HELLO" -C -x c -y c -a c) everything works fine, deleting everything from the display and putting the new text

**Comment by JerwuQu** [comment]
Nice repository and thank you for the bug report!

I will look into it when I have time (hopefully later this week) 😃

**Comment by JerwuQu** [comment]
As for the other things, you mentioned Apex Pro and more text options.

I suspect the keyboard uses the protocol differently and I would not be able to implement it myself without actually having the keyboard.

For text options, I'm fully up for implementing them but I would need to figure out how make the interface good. It might make sense to have scroll options for images too, for example. Should text be scrolling all the time, or only if it covers more than the screen? Should it scroll back and forth, or scroll of screen and appear at the other side, or some other variant? I don't personally require these text options right now so if you could assist with these decisions that would be great :)

**Comment by pascaltuf** [comment]
Hi, sorry I didn't answer you.
As for the space problem, you're right, I added an unnecessary space that caused me that problem, thanks for pointing it out.

For text scrolling, it would be better if:
By default, the program only activates scrolling if it exceeds the display characters (18 characters), but if you want, you could do it in such a way that, as you did to center the text where you put the options, you could force scrolling even to less than 18 characters at the user's discretion.
As for text scrolling, it would be nice if it scrolls like the video (it's the gamesense-essentials program that I showed you in the past, from which I took inspiration to make the script to show the song title)


https://github.com/user-attachments/assets/173a8c44-de93-4f55-a23a-18f346de787b



As for my scripts, have you had a chance to test them?
Do they work well or are there problems?
If you have tested it, can you tell me with which program (Spotify?)
Thanks in advance if you give me some Feedback 😊

**Comment by JerwuQu** [comment]
Hello again!

No problem, and thanks for the example.

I'll try out your scripts later, and give text scrolling a try after I've figured out how it should be implemented.

I'd be interested to know more about gamesense-essentials and if there are things ggoled could do to become better :)

**Comment by pascaltuf** [comment]
gamesense-essentials is program created by mtricht, that works for all steelseries devices that have a screen, you can find it on github :
https://github.com/mtricht/gamesense-essentials

**Comment by JerwuQu** [comment]
Tried your scripts and they work for me with Spotify :)

Will keep the scripts in mind when implementing text scrolling and such. It will be a big refactor since it will require `ggoled` to be able to run in the background to animate the display, but I'm working on it!

**Comment by JerwuQu** [comment]
Hello

Very basic scrolling has been added. This uses the `--scroll` flag (like `--scroll slow`). This required changing how the command works a bit since it needs to stay running to be able to scroll. I also went ahead and changed the defaults for position, alignment, and clearing. You will need to update your scripts.

As example, here's your Time.ps1 script running in stdin mode (meaning it only starts `ggoled` once!):
```ps1
& {
    while ($true) {
        $currentTime = (Get-Date -Format "HH:mm:ss")
        "Time : $currentTime
@"
        Start-Sleep -Seconds 1
    }
} | ggoled text -d "@"
```

This unfortunately requires [PowerShell 7](https://github.com/PowerShell/PowerShell) because [PowerShell 5 is bugged with pipes](https://stackoverflow.com/a/76064200). 

You can still run it like you did before by just calling `ggoled text "Some text"`, but it's more difficult to use for scripts if you want scrolling compared to using the stdin mode.

I will likely improve on the text features in the future but this should be an improvement for now.

**Comment by pascaltuf** [comment]
Thanks for trying my script and I'm glad it works 😊
Thanks again for adding text scrolling and removing the arguments to center the text, so the text is always centered and it's better that way.
The scrolling works perfectly, but now the scripts don't work because the program is still running and it's a bit difficult to make it change the text, if you can make it so that even if the program is running it accepts a new text so it can be changed that would be perfect.
The only thing I could try is to make it so that every time it finds a new text, the script has to close ggoled and run it again with the new text.

**Comment by pascaltuf** [comment]
I was looking to update the script and I noticed this detail that would be useful if it could be changed:
When there are multiple texts, the first text, for example the title of the song, if it does not exceed the screen must remain still and the second text, which can be multiple artists, must move, leaving the first text still. Instead, currently when you activate scrolling, scrolling is activated for all texts regardless of the characters.
For example try this: ggoled text "Somebody (2024)`nGotye, Kimbra, FISHER, Chris Lake, Sante Sansone" --scroll fast

The same in reverse so only if it exceeds the characters you have to activate the scrolling of one of the texts or all 2.

Instead if the scrolling option is active and does not exceed the characters on the screen the scrolling is not activated but the program remains active, and could be modified so that even if the option is active and the characters are all visible on the screen, it does not activate the scrolling option and therefore so that I can send another text again (ggoled closes)

For example with: ggoled text "ABCDEFGHIJKLMNOPQR" --scroll fast

The program remains active but the scrolling is not active, so for example it must completely bypass the scrolling option so ggoled closes, instead with this:

ggoled text "ABCDEFGHIJKLMNOPQRS`nABCDEFGHIJKLMNOPQR" --scroll fast

it has to be enabled because the first text exceeds the characters, the second one does not so the second text must remain still and only the first text must move, as said above
(This rule must work for all possible lines from the display and therefore four not only the first two :-)

I hope you will make these changes if you have time, I am very happy with how this software is getting better every time you add new features

**Comment by JerwuQu** [comment]
> The scrolling works perfectly, but now the scripts don't work because the program is still running and it's a bit difficult to make it change the text, if you can make it so that even if the program is running it accepts a new text so it can be changed that would be perfect.
The only thing I could try is to make it so that every time it finds a new text, the script has to close ggoled and run it again with the new text.

This is what the `-d "@"` argument along with `@` in the text takes care of. It signals to update the screen. Give the script I sent a try! 

---

>  When there are multiple texts, the first text, for example the title of the song, if it does not exceed the screen must remain still and the second text, which can be multiple artists, must move, leaving the first text still. Instead, currently when you activate scrolling, scrolling is activated for all texts regardless of the characters.

Yes, this is something I plan to fix 😄 
To put it another way, it currently scrolls the whole text block rather than individual rows. 

---

> For example with: ggoled text "ABCDEFGHIJKLMNOPQR" --scroll fast
> The program remains active but the scrolling is not active, so for example it must completely bypass the scrolling option so ggoled closes, instead with this:
> ggoled text "ABCDEFGHIJKLMNOPQRS`nABCDEFGHIJKLMNOPQR" --scroll fast
> it has to be enabled because the first text exceeds the characters, the second one does not so the second text must remain still and only the first text must move, as said above
(This rule must work for all possible lines from the display and therefore four not only the first two :-)

I think in this case I actually want the `--scroll` option to take priority so that the program behaves the same everytime in a script. The `--scroll` option is what tells the program to stay running, not the size of the text.

**Comment by pascaltuf** [comment]
I can't run the script you sent above of the time, I also put an output of the time the time in the console shows it to me but the ggoled command doesn't even run it

**Comment by pascaltuf** [comment]
What is the "-d" option for?
Does the @ sign mean it takes this text? :

"Time : $currentTime
@"

**Comment by pascaltuf** [comment]
Ok, now I understand thanks to your script how I can update the screen even with scrolling active

**Comment by pascaltuf** [comment]
Why does scrolling not work when I do this?

"abcdefghijklmnopqrstuvwxyz
@" | ggoled text -d "@" --scroll fast

I'm running this line directly on powershell to understand how it works
Am I doing something wrong?

**Comment by JerwuQu** [comment]
🥳 

> Why does scrolling not work when I do this?

The scrolling stop because stdin is closed, this is by design. 
If you want to persist while using this kind of input you would do `ggoled text --scroll fast "abcdefghijklmnopqrstuvwxyz"` instead.
Though I understand this is perhaps not very obvious... I will consider changing this if that's more obvious.

I'm considering creating an application similar to the gamesense-essentials you linked so most user don't have to worry about writing scripts at all. Would this be interesting?

**Comment by pascaltuf** [comment]
Yes, of course that would be interesting, an app that can run the whole thing in the background that could display the time or the song title would be awesome without using the terminal

**Comment by pascaltuf** [comment]
Does the language you are using allow you to create .exe apps that can run in the background or do you need to switch to another programming language?

**Comment by JerwuQu** [comment]
> Does the language you are using allow you to create .exe apps that can run in the background or do you need to switch to another programming language?

Yes this is possible :)
I'll give it a shot.

**Comment by JerwuQu** [comment]
There's now a `ggoled_app.exe` included in the build that simply shows time and media :)

See if it works for you.

**Comment by pascaltuf** [comment]
Wow, it works great thanks for doing this, it works perfectly.
But in my opinion it could be improved like this:
I noticed that you are taking the title of the song and the artist from windows, doing so automatically any program that uses an audio is shown on the screen, so it works on tidal, spotify, youtube music, soundcloud which is a site, and also non-music videos on youtube, this is actually fine, but for example on tidal not all the artists are shown but only the first, I think it is a windows thing, because even on spotify in case of multiple artists only one artist is shown (the first).
In my opinion if it is possible to do it, you could allow the user to choose it in this way, currently when I right-click on the app in the background only the exit option comes out, here you could add other options such as a list of apps on where to take the song title, which you will take from the title of the .exe app running which as I did on my script is divided by the " - " such as:

Song Title - Artist 1, Artist 2

The same spotify except that spotify I don't know why even if there are more artists it still takes only one artist and the title is backwards:

Artist - Song Title

For spotify you don't need to do this, the Windows one is already fine, because it would give the same result unless you do it like gamesense which I think uses the spotify sdk to take the song but I think it's much more difficult.

If you can make it so that you can show for example a menu from which to take the song that I click on and there are options like:

Windows
Tidal

Tidal when the music is paused comes out written in the title:
"TIDAL" so as in my script if it finds that word or the app is not open it does not show anything otherwise it shows the title and the artists.
The same spotify: "Spotify"

You could also add other options like gamesense type:
Flip song and artist (which swaps the title and the artist, useful if you add spotify among the apps from which to take the title of the song that has the artist first)
Show time (possibility to deactivate the time)
Show song (possibility to deactivate the song and therefore for example you could move the time to the center if the song is deactivated)
Move time to the center (moves the time to the center and if the song is playing it shows it and the time is no longer shown until the song is paused or stopped)

You are doing an incredible job and every time a new update comes out I can't wait to try it, thanks for everything ❤️

**Comment by JerwuQu** [comment]
Hello hello!

There's a new update for the application. I wrote about it in the README if you want to take a look.

Grabbing media info from process window titles is possible but a little tricky, so I will do it later on 👍🏻 

I'm not sure I want to go crazy with options regarding what and how to display things. I think what I have now makes sense for most users.

Thank you a lot for your patience and good feedback, it makes developing this more fun 😊

...since I have now actually implemented text scrolling and font changing, I will move the task about media titles to a separate issue for clearer separation, and close this one :)
